### PR TITLE
Support arbitrary baud rates on BSDs

### LINKED ISSUE #842: Termios API does not allow for getting or setting arbitrary baud rates on BSDs
The BSDs don't need to use the baud rate constants and instead can specify the baud rates directly. Unfortunately this isn't exposed through `nix`'s APIs.

I think the best option here is to change the termios API for the BSDs and add `From<BaudRate> for u32`, then you could call both functions by using `rate.info()`.

### PR DESCRIPTION
**Pull request by Susurrus**
Closes #842

### CODE REVIEW

**Comment by Susurrus** [comment]
It looks like things are still not correct for mac platforms; it needs to be decided if u32 is a fine output type or if it should really be `speed_t`, which is 32 on 32-bit platforms and 64 on 64-bit platforms. Considering the u32 covers up to 4 billion baud in this scenario, I'm going to be it's fine using u32 across all platforms.

**Comment by asomers** [comment]
If you implement `TryFrom<u32>` for `BaudRate`, then it will be easier to write portable code.  The following would work on all platforms:
```rust
cfsetspeed(&mut t, BaudRate::B9600);
let b : BaudRate = cfgetispeed(&t).try_into().unwrap();
```

**Comment by Susurrus** [comment]
TryFrom is nightly-only. I'll add a FIXME to change the API in the future.

**Comment by Susurrus** [comment]
I'll take that as an implicit review @asomers, but cancel this merge if I shouldn't have.

bors r+

**Comment by asomers** [comment]
Not yet!  That's was just what I could come up with while still at work.  I'm going to do the full review now.

bors cancel

File: src/sys/termios.rs [hunk 1]
```diff
@@ -761,22 +889,117 @@ libc_bitflags! {
...
+
+        /// Set output baud rate (see
+        /// [cfsetospeed(3p)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/cfsetospeed.html)).
+        ///
+        /// `cfsetospeed()` sets the output baud rate in the given termios structure.
+        pub fn cfsetospeed(termios: &mut Termios, baud: BaudRate) -> Result<()> {
+            let inner_termios = unsafe { termios.get_libc_termios_mut() };
+            let res = unsafe { libc::cfsetospeed(inner_termios, baud as libc::speed_t) };
+            termios.update_wrapper();
+            Errno::result(res).map(drop)
```
Comment (asomers) [inline src/sys/termios.rs]: Personally, I find `.map(|_| ())` to be more intuitive than `.map(drop)`, though it is a little longer.

File: src/sys/termios.rs [hunk 2]
```diff
@@ -354,6 +475,13 @@ impl From<libc::speed_t> for BaudRate {
     }
 }
 
+// TODO: Include `TryFrom<u32> for BaudRate` once that API stabilizes
+impl From<BaudRate> for u32 {
```
[the remark this answers is not in the document]
Comment (Susurrus) [reply src/sys/termios.rs]: Good catch, I originally meant to only expose this on BSDs.

File: src/sys/termios.rs [hunk 3]
```diff
@@ -761,22 +889,117 @@ libc_bitflags! {
...
-/// [cfgetospeed(3p)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/cfgetospeed.html)).
-///
-/// `cfgetospeed()` extracts the output baud rate from the given Termios structure.
-pub fn cfgetospeed(termios: &Termios) -> BaudRate {
-    let inner_termios = termios.get_libc_termios();
-    unsafe { libc::cfgetospeed(&*inner_termios) }.into()
+        /// Get output baud rate (see
+        /// [cfgetospeed(3p)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/cfgetospeed.html)).
+        ///
+        /// `cfgetospeed()` extracts the output baud rate from the given Termios structure.
```
[the remark this answers is not in the document]
Comment (Susurrus) [reply src/sys/termios.rs]: I don't think it should, Termios is not a struct, `termios` is, Termios is the name of the API.

File: src/sys/termios.rs (hunk 1 above)
Comment (Susurrus) [reply src/sys/termios.rs]: I agree.

File: src/sys/termios.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Susurrus) [reply src/sys/termios.rs]: NVM, I agree with what you're saying.

File: src/sys/termios.rs [hunk 4]
```diff
@@ -761,22 +889,117 @@ libc_bitflags! {
...
+        /// `cfsetospeed()` sets the output baud rate in the given termios structure.
+        pub fn cfsetospeed<T: Into<u32>>(termios: &mut Termios, baud: T) -> Result<()> {
+            let inner_termios = unsafe { termios.get_libc_termios_mut() };
+            let res = unsafe { libc::cfsetospeed(inner_termios, baud.into() as libc::speed_t) };
+            termios.update_wrapper();
+            Errno::result(res).map(drop)
+        }
+
+        /// Set both the input and output baud rates (see
+        /// [termios(3)](http://man7.org/linux/man-pages/man3/termios.3.html)).
```
[the remark this answers is not in the document]
Comment (Susurrus) [reply src/sys/termios.rs]: Indeed, corrected.

**Comment by Susurrus** [comment]
@asomers Take a look-see now, might be more to your liking.

**Pull request merged**